File: lecture012/permutation.java
import java.util.Scanner;
import java.util.ArrayList;

public class permutation {
    public static Scanner scn = new Scanner(System.in);

    public static void main(String[] args) {
        solve();
    }

    public static void solve() {
        System.out.println(permutation("abc"));
    }

    public static ArrayList<String> permutation(String str) {
        if (str.length() == 0) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }
        char ch = str.charAt(0);
        String restStr = str.substring(1);

        ArrayList<String> recAns = permutation(restStr);
        ArrayList<String> myAns = new ArrayList<>();

        for (String s : recAns) {
            for (int i = 0; i <= s.length(); i++) {
                String one = s.substring(0, i);
                String two = s.substring(i);

                String ans = one + ch + two;
                myAns.add(ans);
            }
        }
        return myAns;
    }
}

File: lecture014/permutation.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;
int permutation(string str, string ans)
{
    if (str.length() == 0)
    {
        cout << ans << " " << endl;
        return 1;
    }
    int count = 0;
    for (int i = 0; i < str.length(); i++)
    {
        char ch = str[i];
        string one = str.substr(0, i);
        string two = str.substr(i + 1);

        count += permutation(one + two, ans + ch);
    }
    return count;
}
void solve()
{
    // cout << permutation("abc", "") << " ";
    cout << permutation("aab", "") << " ";
}
int main(int args, char **argv)
{
    solve();
}